Allow some functions to take inline JSON: mapdata, mergedata, and others

tzz opened the pull request:
Related to https://dev.cfengine.com/issues/7871 (when it's implemented, all CFEngine functions that take lists or data containers will be able to use this functionality).

With this patch, many functions will be able to take inline JSON instead of a list name or data container name. We simply try to parse the varname as JSON; since neither `@(listname)` nor `containername` are valid JSON, we end up with a reliable way to parse JSON.

`mergedata()` was a little more complicated because it supports symbolic wrapping, but it's a pretty trivial change otherwise. It also reduces line count, which is a good thing...

Example:
```
bundle agent test
{
  vars:
      "user3" data => data_expand('["$(sys.os)"]');
      "user2" data => mapdata("none", '$(this.v)', '["x"]');
      "user" data => mergedata('[]', '["x"]');
}
```

Result bundle state:

```json
{
  "user": [
    "x"
  ],
  "user2": [
    "x"
  ],
  "user3": [
    "linux"
  ]
}
```

tzz commented:
The second commit here introduces the ability to say

```
      "foo" string => "bar";
      "userfoo" data => data_expand('[foo]');

      "user3" data => data_expand('["$(sys.os)"]');
      "user3a" data => data_expand('[test.user3,]');
```

in other words, to refer to a variable by name inside JSON and have it imported. Unfortunately this breaks at least 5 tests (`01_vars/02_functions/url_get_local.cf` `01_vars/02_functions/network/url_get.cf` `01_vars/02_functions/mergedata_with_list_iteration.cf` `01_vars/02_functions/maparray_multi_index.cf` `01_vars/02_functions/data_readstringarrayidx.cf`)
so I need some feedback:

* is it useful? @nickanderson ?
* do the test failures look fixable?
* is anyone interested in helping out with the work? @worr @kacf @jimis ?

If the second commit is not useful, I'll drop it.

tzz commented:
Full test I used for the second commit, all the references were expanded correctly:
```
bundle agent main
{
  methods:
      "test";

  vars:
      "test_state" data => bundlestate(test);
      "test_string" string => storejson(test_state);

  reports:
      "$(this.bundle): state of things = $(test_string)";
}

bundle agent test
{
  vars:
      "foo" string => "bar";
      "userA" data => data_expand('{ "x": test.user3 }');
      "user9" data => data_expand('{ "x": user3 }');
      "user8" data => data_expand('[user3]');
      "user7" data => data_expand('{foo:user3}');
      "user6" data => data_expand('[foo]');
      "user5" data => data_expand('[test.user3,]');
      "user4" data => data_expand('[test.user3]');
      "user3" data => data_expand('["$(sys.os)"]');
      "user2" data => mapdata("none", '$(this.v)', '["x"]');
      "user" data => mergedata('[]', '["x"]');
}
```

kacf commented:
In general I think this is very good idea, we just need to be careful we don't introduce any ambiguity.

> in other words, to refer to a variable by name inside JSON and have it imported.

I'm not quite clear on this; how do you differentiate between a variable to be expanded and a simple string. Is it just that it's unquoted?

How will it work with function returns? Can a function returning a container be used as an argument to a different function? I suppose they would be different types, but this might not be that hard to account for.

I haven't looked at the test failures yet. Do you see any recurring pattern?

tzz commented:
> I'm not quite clear on this; how do you differentiate between a variable to be expanded and a simple string. Is it just that it's unquoted?

At the policy level, if a function expects a container name, we also allow strings that start with `{` or `[` because those can't be container names.

Inside the JSON, currently we reject bare words unless they are keys. So we'll now allow bare words such as `myvar` or `mybundle.myvar`. Except for bareword keys, this is fully consistent.

> How will it work with function returns? Can a function returning a container be used as an argument to a different function? I suppose they would be different types, but this might not be that hard to account for.

I don't think it helps in that case currently, but we could hack it later (as the same new function to handle var refs OR inline JSON we introduce here becomes standard as in https://dev.cfengine.com/issues/7871, it will be easier to add new behavior in a single place).

> I haven't looked at the test failures yet. Do you see any recurring pattern?

I think it's either a bug, or something with bareword keys. I'll keep digging if it's not visible to anyone else...

nickanderson commented:
I can see how it could be useful in some cases, but I can't think of how I would use it right now.

tzz commented:
Hacking around `json.c` makes me think of how much easier life would be if we used libyaml to parse JSON.

tzz commented:
OK, all the tests pass, but I had to use a regex to capture the possible variations of `varname`, `bundle.varname`, and `varname[lookup]` which will slow down this specific code path a little bit. I wrote a new test to ensure all of that is checked.

This is ready for review...

kacf commented:
> Inside the JSON, currently we reject bare words unless they are keys. So we'll now allow bare words such as `myvar` or `mybundle.myvar`. Except for bareword keys, this is fully consistent.

Consistent perhaps, but not intuitive nor easy to remember. This means that a bareword in a key and a value will have very different and conflicting meanings. I'm wondering if variable support is demanding a bit much of this feature. I think we at least need to enforce "variable syntax" on it to make sure it stays readable for novice policy readers. So perhaps we could stay with barewords as you originally proposed, but we require `$(var)` around the name, like in regular policy. What do you think?

Another, more technical concern is how this works out if we indeed choose to use libyaml to parse JSON one day. Will it accept barewords in JSON values at all? It might make it more difficult to integrate with libyaml later.

And what about typing YAML directly into `mergedata`?

worr commented:
Should we accept barewords at all? They aren't allowed by the JSON spec, so imo allowing them at all anywhere in any JSON data container is a bug.

Like @kacf mentioned, this will likely make it harder to switch to a different parser later.

tzz commented:
> Consistent perhaps, but not intuitive nor easy to remember. This means that a bareword in a key and a value will have very different and conflicting meanings. I'm wondering if variable support is demanding a bit much of this feature. I think we at least need to enforce "variable syntax" on it to make sure it stays readable for novice policy readers. So perhaps we could stay with barewords as you originally proposed, but we require $(var) around the name, like in regular policy. What do you think?

I don't think it would benefit CFEngine to grow its own way to specify inline JSON, and this PR is all about ease of use. I'd rather stay consistent with Javascript and use barewords.

Remember that these are strings, so the evaluator would see `$(var)` and do nasty things to it. We'd need true JSON markup at the parser level to avoid that (and that would make it clear that these are not just strings). Maybe a special marker like `inlinejson:{...}` instead of putting it in a string?

> Another, more technical concern is how this works out if we indeed choose to use libyaml to parse JSON one day. Will it accept barewords in JSON values at all? It might make it more difficult to integrate with libyaml later.

True. Actually researching into libyaml I found it's not good at parsing pure JSON anyway, it's too strict in every way. We'll probably want libjson or something like it.

> And what about typing YAML directly into mergedata?

Because YAML, unlike JSON, is an ambiguous format, I wouldn't allow it.

> Should we accept barewords at all? They aren't allowed by the JSON spec, so imo allowing them at all anywhere in any JSON data container is a bug.

CFEngine and Javascript allow bareword keys in JSON already for key values, see https://github.com/cfengine/core/pull/2247 and especially https://github.com/cfengine/core/pull/2247#issuecomment-116605655

kacf commented:
> > Consistent perhaps, but not intuitive nor easy to remember. This means that a bareword in a key and a value will have very different and conflicting meanings. I'm wondering if variable support is demanding a bit much of this feature. I think we at least need to enforce "variable syntax" on it to make sure it stays readable for novice policy readers. So perhaps we could stay with barewords as you originally proposed, but we require $(var) around the name, like in regular policy. What do you think?
> 
> I don't think it would benefit CFEngine to grow its own way to specify inline JSON, and this PR is all about ease of use. I'd rather stay consistent with Javascript and use barewords.
> 
> Remember that these are strings, so the evaluator would see `$(var)` and do nasty things to it. We'd need true JSON markup at the parser level to avoid that (and that would make it clear that these are not just strings). Maybe a special marker like `inlinejson:{...}` instead of putting it in a string?

Isn't using barewords in JSON values *already* growing a custom way of parsing JSON? I don't think CFEngine would benefit from that either. Perhaps you should ping me in the chat later. I'm not sure I fully understood this.

> Actually researching into libyaml I found it's not good at parsing pure JSON anyway, it's too strict in every way. We'll probably want libjson or something like it.

I'd be interested in knowing more details about this. Were there valid JSON files that would not be parsed by libyaml? Or are we already too relaxed in our own parser that we have introduced incompatibilities?

> > And what about typing YAML directly into mergedata?
> 
> Because YAML, unlike JSON, is an ambiguous format, I wouldn't allow it.

Gotcha.

> > Should we accept barewords at all? They aren't allowed by the JSON spec, so imo allowing them at all anywhere in any JSON data container is a bug.
> 
> CFEngine and Javascript allow bareword keys in JSON already for key values, see https://github.com/cfengine/core/pull/2247 and especially https://github.com/cfengine/core/pull/2247#issuecomment-116605655

Indeed, incompatible with the full strictness of the spec, but "reasonable" in the sense that it seems to be de facto supported.

tzz commented:
OK, rebased against latest master.

kacf commented:
Thanks!

I won't post the whole log, but the summary from the IRC discussion is that bareword values as variables are already supported when using inline JSON in `mergedata`, so there is little point in debating whether we should use them now. Better to stay consistent and extend the feature here.

The point about libyaml/libjson is unlikely to be relevant, because libyaml is *too*strict for JSON parsing, and libjson doesn't give us much benefit over what we already have.

I'll take a more detailed look at the code shortly.

kacf commented on libutils/json.h:
```diff
@@ -354,13 +354,19 @@ JsonElement *JsonArrayGet(const JsonElement *array, size_t index);
+  libpromises into libutil
   */
-JsonParseError JsonParse(const char **data, JsonElement **json_out);
+JsonParseError JsonParse(void *ctx, JsonLookup *lookup, const char **data, JsonElement **json_out);
```
Can you please leave the `JsonParse` function as is, and make a new function `JsonParseWithLookup`? The former can call the latter with null arguments. Then we don't have to change it everywhere, and the interface stays clean.

Also, slightly more cosmetic, but I think `ctx` should be renamed to `map`, since in principle it has no relation to `ctx`, it can be anything. Same thing for the `JsonLookup` function above. The comment about `ctx` being used as the map can then be moved to libpromises, at the place where it is actually passed in, which is where this information belongs. The information about the `map` argument can simply be something like `map which the JsonLookup function will use to look up references`, or something like that.

In general there shouldn't be any references to ctx in libutils, neither comments nor otherwise, since it is supposed to be independent from it.

tzz replied:
OK; fixed. I renamed `ctx` to `lookup_context` because I think it's the best name for it, and `lookup` is now `lookup_function`. I hope that's acceptable, and you're right that the change is much smaller and cleaner that way.

kacf replied:
Yep, that's good!

kacf commented on libpromises/evalfunction.c:
```diff
@@ -195,8 +196,9 @@ static VarRef* ResolveAndQualifyVarName(const FnCall *fp, const char *varname)
-                                      const DataType disallowed_datatypes[], size_t disallowed_count)
+static JsonElement* VarRefValueToJsonAllowingScalars(EvalContext *ctx, const FnCall *fp, const VarRef *ref,
+                                                     const DataType disallowed_datatypes[], size_t disallowed_count,
+                                                     bool allow_scalars)
```
Why can't `disallowed_datatypes` be used here instead of introducing yet another argument `allow_scalars`? Can't you just put the scalar type as disallowed in the appropriate places?

tzz replied:
Because we disallow scalars by default, unfortunately.

kacf replied:
Ah ok, I see.

kacf commented on libpromises/evalfunction.c:
```diff
@@ -313,13 +328,85 @@ static JsonElement* VarRefValueToJson(EvalContext *ctx, const FnCall *fp, const
+    else
+    {
+        return inline_data;
+    }
```
Weird code flow with empty if. I propose to change it to the following:

```
    if (res == JSON_PARSE_OK)
    {
        if (JsonGetElementType(inline_data) == JSON_ELEMENT_TYPE_PRIMITIVE)
        {
            JsonDestroy(inline_data);
            inline_data = NULL;
        }
        else
        {
            return inline_data;
        }
    }
```

tzz replied:
Thanks for catching that. Fixed.

kacf commented on libpromises/evalfunction.c:
```diff
@@ -2971,91 +3046,20 @@ static FnCallResult FnCallMergeData(EvalContext *ctx, ARG_UNUSED const Policy *p
-            return FnFailure();
-        }
-
-        JsonElement *json = VarRefValueToJson(ctx, fp, ref, NULL, 0);
```
+1. Nice to have this gathered in one place!

kacf commented on libpromises/evalfunction.c:
```diff
@@ -251,7 +264,8 @@ static JsonElement* VarRefValueToJson(EvalContext *ctx, const FnCall *fp, const
                     {
-                        Log(LOG_LEVEL_DEBUG, "%s: got ref with starting depth %zd and index count %zd", fp->name, index_offset, var->ref->num_indices);
+                        Log(LOG_LEVEL_DEBUG, "%s: got ref with starting depth %zd and index count %zd",
+                            fp ? fp->name : "VarRefValueToJson", index_offset, var->ref->num_indices);
```
Can you make this `fp` check at the top of the function instead, and store the name in a different variable. Then you don't need to check this every place where it's used.

Same goes for several uses below this one.

Alternatively, since `fp` is only used to fetch the name AFAICS, perhaps you shouldn't pass in a null pointer at all, but you should change it to a string and pass in `"LookupVarRefToJson"` from the one function that now passes null. It still won't be fully verbose in the sense that the real calling function from the policy is masked, but it would take quite a bit of data passing to get that information all the way here.

tzz replied:
I made a variable `fp_name` because, the way we use `fp`, I think it may be needed in the future so just passing a string is not future-proof. I hope that's OK.

kacf replied:
Sure!

tzz commented:
Thanks for the review. I've fixed all but the "disallow scalar" issue, and rebased against master. The change is much smaller thanks to the `JsonParseWithLookup` suggestion.

kacf commented:
trigger build

kacf commented:
Build is green for 5cce98b! I assume that this should have a mention in the docs as well. Something similar to https://github.com/cfengine/documentation/commit/ef825dd14b57b61a1a17928bbd2e353164eaf76f perhaps?

tzz commented:
docs in https://github.com/cfengine/documentation/pull/1306 (with some notes)

kacf commented:
Thanks! Both merged, this one manually.

tzz commented:
CHEERS
